### ISSUE TITLE
Option to build into temp directory before moving to output directory

### ISSUE THREAD

**Issue by SchnWalter** [ng-packagr/ng-packagr#2022]
## Type of Issue

```
[ ] Bug Report
[x] Feature Request
```

## Description

It would be nice to have an option to move the compiled output from a temporary directory instead of building directly into the output directory, this is especially useful when `npm link` is used to link the built library into another project.

On large projects with a lot of sub-packages, the `--watch` rebuild can be slow if there are many dependencies, this causes the `ng serve` from the project consuming the library to fail because of missing sub-packages (which are yet to be rebuilt). Adding an option to move the compiled output into place would avoid all these issues caused by slower builds.

#### How To Reproduce

* run `ng build --configuration=development --watch` in a very large library project, with build time longer than 20-30 seconds.
* run `npm link` to link the build into an Angular application. And make sure you have `preserveSymlinks` set in `angular.json` and `tsconfig.json`.
* run `ng serve --configuration=development` for the application.
* Trigger a slower rebuild in the library and wait, during the rebuild, the application dev server will see the first change and start rebuilding the application causing it to fail because of missing library sub-packages, which are yet to be rebuilt by `ng-packagr`.

Reproduction: [2022-slow-rebuild-watch-error](https://github.com/SchnWalter/repo-for-reproducing-angular-issues/blob/2022-slow-rebuild-watch-error/README.md)

#### Expected Behaviour

The application build should succeed.

#### Version Information

```bash
# node --version
v12.22.1
# npm --version
6.14.13
# ./node_modules/.bin/ng-packagr --version
ng-packagr:            12.1.0
@angular/compiler:     12.1.0
rollup:                2.52.3
typescript:            4.3.4
```

P.S. Depending on the time I have, I will probably take a look at this in the next couple of weeks, but I wanted to create an issue so that people know about this.

**Comment by alan-agius4** [comment]
Hi, 

I am not so keen into creating a temporary directory and move the files around during a rebuilds. We should reduce IO during rebuilds, although with a copy-on-write reflink (COW) it shouldn't be that bad. Although a possible caviat of this would be that assets would end up truly copied. Since they are already being copied using a COW strategy and a copy of a COW will likely result in a true copy of the file. Will likely need some proper testing and debugging if we go with the temporary directory solution.

I think what is happening here, is that with Webpack 5, node_modules are immutable by default meaning that a standard NPM link will not work. 

Recently we introduced https://github.com/ng-packagr/ng-packagr/commit/4e13041a7cc47918dcecd4dd94f2f57edc0e747c which changes the package version during rebuilds. The problem here is that the primary entry-point is typically built first which means that the `version` get updated too soon. 

What is concerning however is your following comment. 
> Trigger a slower rebuild in the library and wait, during the rebuild, the application dev server will see the first change and start rebuilding the application causing it to fail because of missing library sub-packages, which are yet to be rebuilt by ng-packagr.

I am not sure what is happening here, since files and sub-packages are not deleted during rebuilds.

**Comment by SchnWalter** [comment]
I'll try to reproduce this in a generate "big" library and see if I can reproduce the "missing import error" (or however it was named).

**Comment by SchnWalter** [comment]
I managed to trigger the same error in a fresh project, but in my initial report, the error appears while on `ng build --watch + ng serve`. In a [freshly generated project](https://github.com/SchnWalter/repo-for-reproducing-angular-issues/blob/2022-slow-rebuild-watch-error/README.md), this happens only after killing and restarting the library `ng build --watch` command, while the application `ng serve` is still active. And to recover from this state, you need to kill and restart `ng serve`.

I'll keep digging to see what's happening in my original application, though this "kill + watch" situation might also be something that we could address.

P.S. There are no "delete output" flags set in the project that might clean some directories during builds.

**Comment by SchnWalter** [comment]
I saw https://github.com/ng-packagr/ng-packagr/commit/4e13041a7cc47918dcecd4dd94f2f57edc0e747c, that's when I decided to (re)investigate this issue. Until now we've wrapped the ngPackagr build around our own command to cleanup and copy after build and it worked without symbolic links, but this was slow.

I didn't spend to much time on fixing that because I was under the impression that `npm link` is something frowned upon in the angular community, at least this was my impression from the Angular CLI issue queue and I thought that it's futile to "go against the stream".

**Comment by SchnWalter** [comment]
> What is concerning however is your following comment.
> 
> > Trigger a slower rebuild in the library and wait, during the rebuild, the application dev server will see the first change and start rebuilding the application causing it to fail because of missing library sub-packages, which are yet to be rebuilt by ng-packagr.
> 
> I am not sure what is happening here, since files and sub-packages are not deleted during rebuilds.

It turns out that in our update to Angular v12 we also took advantage of the new "version stamping" (https://github.com/ng-packagr/ng-packagr/commit/4e13041a7cc47918dcecd4dd94f2f57edc0e747c) and we switched to using `npm link` instead of a custom build script. But our `ng-package.json` file from that branch had an extra flag enabled: `"deleteDestPath": true`, which was causing the error. With this flag disabled, the errors went away.

So, for those reading this at a later time: we use different repositories containing Angular CLI workspaces, each workspace contains or more libraries, all of them with sub-packages and quite a large dependency tree. Some are "core libraries" and we have others that depend on them, and in the end we build these into an application. Because of the "deleteDestPath" in one of our "core" libraries, the `watch` or `serve` commands from some of the workspaces were failing while a core library was being rebuilt. This happens because the `package.json` files from symlink-ed libraries were being deleted and the Angular application builder doesn't like this and it's not "restarting" once the build is finished. The solution was to not clear the destination path, so I will close this issue.

Thanks for your help and patience.
